src/fsharp/FSharp.Core/resumable.fsi

```diff
@@ -175,4 +175,40 @@ module StateMachineHelpers =
         afterCode: AfterCode<'Data, 'Result> 
             -> 'Result
 
+/// <summary>Adding this attribute to the method adjusts the processing of some generic methods
+/// during overload resolution.</summary>
+///
+/// <remarks>During overload resolution, caller arguments are matched with called arguments
+/// to extract type information. By default, when the caller argument type is unconstrained (for example
+/// a simple value <c>x</c> without known type information), and a method qualifies for
+/// lambda constraint propagation, then member trait constraints from a method overload
+/// are eagerly applied to the caller argument type. This causes that overload to be preferred,
+/// regardless of other method overload resolution rules. Using this attribute suppresses this behaviour. 
+/// </remarks>
+///
+/// <example>
+/// Consider the following overloads:
+/// <code>
+/// type OverloadsWithSrtp() =
+///     [&lt;NoEagerConstraintApplicationAttribute&gt;]
```

Comment:
I see, - so in order to get the desired behavior, you just use the attribute on a SRTP function.

Reply:
is this generally-desired behavior? i.e. should we have guidance to library authors that says 'if you're using SRTP with overloading, you _should_ apply NoEagerConstraintApplicationAttribute', in the RFC sense of _should_?

Reply:
This will need to be assessed on a case by case basis.  It depends very much what the intention of the sets of overloads is.

Looking at Ply (and I assume some other SRTP code), there seems to be an intent where, if there is no nominal type information on the caller side, the SRTP overload kicks in and the type inference proceeds.  

For FSharp.Core tasks, the intention is to give the `Task<_>` overload priority.

I think we'd need a longer blog post or article discussing how to design overload sets in F# that include some SRTP constrained overloads and carefully discuss the ramifications of the different choices with many examples.

Reply:
I'm not sure I understand this part: `the SRTP overload kicks in and the type inference proceeds`, how does - in the case of Ply and Taskbuilder - resolving to the explicit `Async<'T>` overload imply the SRTP overload kicked in or was somehow involved? 

> I think we'd need a longer blog post or article discussing how to design overload sets in F# that include some SRTP constrained overloads and carefully discuss the ramifications of the different choices with many examples.

@dsyme I would very much like to work on that! I find the interplay of the pieces involved in overload resolution to be hard to intuit. And with the lack of in depth written material it's very opaque short of tracing the compiler code.

Reply:
> I'm not sure I understand this part: the SRTP overload kicks in and the type inference proceeds, how does - in the case of Ply and Taskbuilder - resolving to the explicit Async<'T> overload imply the SRTP overload kicked in or was somehow involved?

Without the attribute, if the caller side has a nominal known type `Async<'T>` or indeed any type that conflicts with the application of the SRTP constraints, then the SRTP overload constraints won't get eagerly applied and won't be involved.

The problem is always in the cases where there is not enough information on the caller side to rule out the SRTP overload.  In that case the SRTP overload constraints get eagerly applied.   It's a by-product of the pre-filtering (lambda propagation) rule that is "leaky" - the rule is meant to assess the overloads for safe information propagation, without affecting the caller side except in a totally safe way, using information common across all overloads. But in the case of SRTP overloads the rule is "leaky", and the caller side gets SRTP cosntraints applied.